# AfterWork Data Science: Data Wrangling with Python Project

## 1. Defining the Question

### a) Specifying the Data Analysis Question

You need to answer the following guiding questions as part of your analysis. Feel free to come
up with many more other questions.
1. Which bus companies had the highest breakdowns?
2. What were the top 3 reasons for bus delays?
3. How many students were on the buses when they broke down?
4. Which were the most frequent reasons for bus breakdowns?
5. What were the most frequent reasons for the bus running late?
6. What was the average delay time of each reason type?

### b) Defining the Metric for Success

How will you know that your solution will have satisfied your research question?

### c) Understanding the context 

The Bus Breakdown and Delay system collects information from school bus vendors operating out in the field in real time. Bus staff that encounter delays during the route are instructed to radio the dispatcher at the bus vendor’s central office. The bus vendor staff are then instructed to log into the Bus Breakdown and Delay system to record the event and notify OPT. OPT customer service agents use this system to inform parents who call with questions regarding bus service. The Bus Breakdown and Delay system is publicly accessible and contains real time updates. All information in the system is entered by school bus vendor staff.

### d) Recording the Experimental Design

Describe the steps/approach that you will use to answer the given question.

### e) Data Relevance

How relevant was the provided data?

## 2. Reading the Data

In [49]:
# Importing our libraries
# ---
#
import pandas as pd
import numpy as np

In [3]:
# Load the data below
# --- 
# Dataset url = https://bit.ly/BusBreakdownDataset
# --- 
# 

bus_df = pd.read_csv('https://bit.ly/BusBreakdownDataset')

In [4]:
# Checking the first 5 rows of data
# ---
#

bus_df.head(5)

School_Year  Busbreakdown_ID           Run_Type Bus_No Route_Number  \
0   2015-2016          1227538  Special Ed AM Run   2621         J711   
1   2015-2016          1227539  Special Ed AM Run   1260         M351   
2   2015-2016          1227540           Pre-K/EI    418            3   
3   2015-2016          1227541  Special Ed AM Run   4522         M271   
4   2015-2016          1227542  Special Ed AM Run   3124         M373   

          Reason Schools_Serviced             Occurred_On  \
0  Heavy Traffic            75003  11/05/2015 08:10:00 AM   
1  Heavy Traffic            06716  11/05/2015 08:10:00 AM   
2  Heavy Traffic             C445  11/05/2015 08:09:00 AM   
3  Heavy Traffic            02699  11/05/2015 08:12:00 AM   
4  Heavy Traffic            02116  11/05/2015 08:13:00 AM   

               Created_On        Boro  ... How_Long_Delayed  \
0  11/05/2015 08:12:00 AM  New Jersey  ...              NaN   
1  11/05/2015 08:12:00 AM   Manhattan  ...            20MNS   
2  11/05/2015 08:13:00 AM       Bronx  ...            15MIN   
3  11/05/2015 08:14:00 AM   Manhattan  ...           15 MIN   
4  11/05/2015 08:14:00 AM   Manhattan  ...              NaN   

  Number_Of_Students_On_The_Bus  Has_Contractor_Notified_Schools  \
0                            11                              Yes   
1                             2                              Yes   
2                             8                              Yes   
3                             6                               No   
4                             6                               No   

  Has_Contractor_Notified_Parents Have_You_Alerted_OPT  \
0                              No                  Yes   
1                             Yes                   No   
2                             Yes                  Yes   
3                              No                   No   
4                              No                   No   

              Informed_On Incident_Number         Last_Updated_On  \
0  11/05/2015 08:12:00 AM             NaN  11/05/2015 08:12:14 AM   
1  11/05/2015 08:12:00 AM             NaN  11/05/2015 08:13:34 AM   
2  11/05/2015 08:13:00 AM             NaN  11/05/2015 08:13:22 AM   
3  11/05/2015 08:14:00 AM             NaN  11/05/2015 08:14:04 AM   
4  11/05/2015 08:14:00 AM             NaN  11/05/2015 08:14:08 AM   

  Breakdown_or_Running_Late School_Age_or_PreK  
0              Running Late         School-Age  
1              Running Late         School-Age  
2              Running Late              Pre-K  
3              Running Late         School-Age  
4              Running Late         School-Age  

[5 rows x 21 columns]

In [5]:
# Checking the last 5 rows of data
# ---
#
bus_df.tail(5)

School_Year  Busbreakdown_ID             Run_Type  Bus_No Route_Number  \
281105   2016-2017          1338452             Pre-K/EI    9345            2   
281106   2016-2017          1341521             Pre-K/EI    0001            5   
281107   2016-2017          1353044    Special Ed PM Run  GC0112         X928   
281108   2016-2017          1353045    Special Ed PM Run   5525D         Q920   
281109   2016-2017          1353046  Project Read PM Run    2530         K617   

               Reason Schools_Serviced             Occurred_On  \
281105  Heavy Traffic             C530  04/05/2017 08:00:00 AM   
281106  Heavy Traffic             C579  04/24/2017 07:42:00 AM   
281107  Heavy Traffic            09003  05/25/2017 04:22:00 PM   
281108    Won`t Start            24457  05/25/2017 04:27:00 PM   
281109          Other            21436  05/25/2017 04:36:00 PM   

                    Created_On      Boro  ... How_Long_Delayed  \
281105  04/05/2017 08:10:00 AM     Bronx  ...            15-20   
281106  04/24/2017 07:44:00 AM     Bronx  ...          20 MINS   
281107  05/25/2017 04:28:00 PM     Bronx  ...        20-25MINS   
281108  05/25/2017 04:30:00 PM    Queens  ...              NaN   
281109  05/25/2017 04:37:00 PM  Brooklyn  ...            45min   

       Number_Of_Students_On_The_Bus  Has_Contractor_Notified_Schools  \
281105                             7                              Yes   
281106                             0                              Yes   
281107                             0                              Yes   
281108                             0                              Yes   
281109                             7                              Yes   

       Has_Contractor_Notified_Parents Have_You_Alerted_OPT  \
281105                             Yes                   No   
281106                             Yes                   No   
281107                             Yes                  Yes   
281108                             Yes                   No   
281109                             Yes                  Yes   

                   Informed_On Incident_Number         Last_Updated_On  \
281105  04/05/2017 08:10:00 AM             NaN  04/05/2017 08:10:15 AM   
281106  04/24/2017 07:44:00 AM             NaN  04/24/2017 07:44:15 AM   
281107  05/25/2017 04:28:00 PM        90323827  05/25/2017 04:34:36 PM   
281108  05/25/2017 04:30:00 PM             NaN  05/25/2017 04:30:07 PM   
281109  05/25/2017 04:37:00 PM             NaN  05/25/2017 04:37:37 PM   

       Breakdown_or_Running_Late School_Age_or_PreK  
281105              Running Late              Pre-K  
281106              Running Late              Pre-K  
281107              Running Late         School-Age  
281108                 Breakdown         School-Age  
281109              Running Late         School-Age  

[5 rows x 21 columns]

In [6]:
# Sample 10 rows of data
# ---
#
bus_df.sample(n=10)

School_Year  Busbreakdown_ID           Run_Type  Bus_No Route_Number  \
159246   2015-2016          1256153           Pre-K/EI     101            6   
58348    2016-2017          1303405  General Ed AM Run   10491        K1971   
39754    2016-2017          1280555  Special Ed AM Run     530         Q836   
220095   2018-2019          1475566  Special Ed AM Run    3647         K479   
234252   2018-2019          1464477  Special Ed AM Run    8023         M549   
261330   2017-2018          1425651  Special Ed AM Run   16422         X099   
247321   2016-2017          1357386  Special Ed AM Run  FR1630         M114   
230595   2017-2018          1411932  Special Ed AM Run    2205         L737   
39569    2016-2017          1280314  Special Ed AM Run    132D         L799   
250919   2015-2016          1226557  Special Ed PM Run  TN0423         M325   

               Reason   Schools_Serviced             Occurred_On  \
159246  Heavy Traffic               C586  03/15/2016 07:57:00 AM   
58348           Other              22326  11/30/2016 06:15:00 AM   
39754           Other              27466  08/03/2016 07:07:00 AM   
220095  Heavy Traffic  18003,18114,18272  10/24/2018 05:59:00 AM   
234252  Heavy Traffic  05085,06100,06101  09/27/2018 06:20:00 AM   
261330          Other        08487,08688  04/06/2018 07:00:00 AM   
247321  Heavy Traffic              03417  06/13/2017 08:19:00 AM   
230595  Heavy Traffic  13508,13758,17009  02/08/2018 07:39:00 AM   
39569   Heavy Traffic              13414  08/01/2016 07:54:00 AM   
250919  Heavy Traffic              02566  10/30/2015 04:30:00 PM   

                    Created_On       Boro  ... How_Long_Delayed  \
159246  03/15/2016 08:00:00 AM      Bronx  ...            15MIN   
58348   11/30/2016 06:28:00 AM   Brooklyn  ...           15 min   
39754   08/03/2016 07:08:00 AM     Queens  ...           30 MIN   
220095  10/24/2018 06:02:00 AM   Brooklyn  ...        31-45 Min   
234252  09/27/2018 06:29:00 AM  Manhattan  ...        16-30 Min   
261330  04/06/2018 07:26:00 AM      Bronx  ...        16-30 Min   
247321  06/13/2017 08:20:00 AM  Manhattan  ...           25 min   
230595  02/08/2018 07:40:00 AM        NaN  ...        16-30 Min   
39569   08/01/2016 07:56:00 AM     Queens  ...              NaN   
250919  10/30/2015 06:08:00 PM  Manhattan  ...           1 HOUR   

       Number_Of_Students_On_The_Bus  Has_Contractor_Notified_Schools  \
159246                             9                              Yes   
58348                              0                              Yes   
39754                              1                              Yes   
220095                             0                              Yes   
234252                             0                              Yes   
261330                             0                              Yes   
247321                             3                              Yes   
230595                             0                              Yes   
39569                              0                              Yes   
250919                             9                               No   

       Has_Contractor_Notified_Parents Have_You_Alerted_OPT  \
159246                             Yes                  Yes   
58348                               No                   No   
39754                              Yes                   No   
220095                             Yes                  Yes   
234252                             Yes                   No   
261330                             Yes                   No   
247321                             Yes                  Yes   
230595                             Yes                  Yes   
39569                              Yes                   No   
250919                             Yes                   No   

                   Informed_On Incident_Number         Last_Updated_On  \
159246  03/15/2016 08:00:00 AM             NaN  03/15/2016 08:00:37 AM   
5834

In [7]:
# Checking number of rows and columns
# ---
#  
bus_df.shape

(281110, 21)

In [8]:
# Checking datatypes
# ---
# 
bus_df.dtypes

School_Year                        object
Busbreakdown_ID                     int64
Run_Type                           object
Bus_No                             object
Route_Number                       object
Reason                             object
Schools_Serviced                   object
Occurred_On                        object
Created_On                         object
Boro                               object
Bus_Company_Name                   object
How_Long_Delayed                   object
Number_Of_Students_On_The_Bus       int64
Has_Contractor_Notified_Schools    object
Has_Contractor_Notified_Parents    object
Have_You_Alerted_OPT               object
Informed_On                        object
Incident_Number                    object
Last_Updated_On                    object
Breakdown_or_Running_Late          object
School_Age_or_PreK                 object
dtype: object

Record your observations below:

*   Our data has 21 attributes and 281110 records
*   Observation 2



## 3. External Data Source Validation

You can make sure your data matches other credible information. This ensures that the measurements are roughly in line with what they should be and it serves as a check on what other things might be wrong in your dataset.

External validation can often be as simple as checking your data against some other data i.e. country population your dataset with country population on world bank data.

Replace the text found on this text cell with your external data source validation information.

## 4. Data Preparation

### Performing Data Cleaning

In [9]:
# Checking datatypes and missing entries of all the variables
# ---
# 
bus_df.isna().sum()

School_Year                             0
Busbreakdown_ID                         0
Run_Type                                3
Bus_No                                  9
Route_Number                            7
Reason                                  2
Schools_Serviced                        7
Occurred_On                             0
Created_On                              0
Boro                                13461
Bus_Company_Name                        0
How_Long_Delayed                    35608
Number_Of_Students_On_The_Bus           0
Has_Contractor_Notified_Schools         0
Has_Contractor_Notified_Parents         0
Have_You_Alerted_OPT                    0
Informed_On                             0
Incident_Number                    271627
Last_Updated_On                         0
Breakdown_or_Running_Late               0
School_Age_or_PreK                      0
dtype: int64

**We observe the following from our dataset:**
*   We do not have any record with all missing values



In [10]:
# Standardizing your dataset i.e. variable renaming
# ---
#convering the columns to lower case and removing the leading and trailing whitespaces if any

bus_df.columns= bus_df.columns.str.strip().str.lower()
bus_df.dtypes

school_year                        object
busbreakdown_id                     int64
run_type                           object
bus_no                             object
route_number                       object
reason                             object
schools_serviced                   object
occurred_on                        object
created_on                         object
boro                               object
bus_company_name                   object
how_long_delayed                   object
number_of_students_on_the_bus       int64
has_contractor_notified_schools    object
has_contractor_notified_parents    object
have_you_alerted_opt               object
informed_on                        object
incident_number                    object
last_updated_on                    object
breakdown_or_running_late          object
school_age_or_prek                 object
dtype: object

**We observe the following from our dataset:**

*  We have converted our colums to be in lower cases and each word separated by an underscore



In [11]:
# Checking how many duplicate rows are there in the data
# ---
# 
bus_df.duplicated().sum()

0

**We observe the following from our dataset:**

*   No duplicates on our dataset



In [12]:
# Checking if any of the columns are all null
# ---
# 
bus_df.isnull().all().sum()

0

**We observe the following from our dataset:**

*   None of the columns are all null

In [28]:
# Checking if any of the rows are all null
# ---
#
bus_df.isna().sum()

school_year                             0
busbreakdown_id                         0
run_type                                3
bus_no                                  9
route_number                            7
reason                                  2
schools_serviced                        7
occurred_on                             0
created_on                              0
boro                                13461
bus_company_name                        0
how_long_delayed                    35608
number_of_students_on_the_bus           0
has_contractor_notified_schools         0
has_contractor_notified_parents         0
have_you_alerted_opt                    0
informed_on                             0
incident_number                    271627
last_updated_on                         0
breakdown_or_running_late               0
school_age_or_prek                      0
dtype: int64

**We observe the following from our dataset:**

* No any row that has all values as null



In [16]:
# Checking if the "Yes/No" fields contain only these 2 values
# for have_you_alerted_opt variable
# ---
# Hint: Use unique() function
#
list(bus_df['have_you_alerted_opt'].unique())

['Yes', 'No']

**We observe the following from our dataset:**

*   The `have_you_alerted_opt` variable has only `['Yes', 'No']` values



In [17]:
# Checking if the "Yes/No" fields contain only these 2 values
# for has_contractor_notified_parents variable
# ---
# 
list(bus_df['has_contractor_notified_parents'].unique())

['No', 'Yes']

**We observe the following from our dataset:**

*   The `has_contractor_notified_parents` variable has only `['Yes', 'No']` values



In [18]:
# Checking if the "Yes/No" fields contain only these 2 values
# for has_contractor_notified_schools variable
# ---
#
list(bus_df['has_contractor_notified_schools'].unique())  

['Yes', 'No']

**We observe the following from our dataset:**

*   The `has_contractor_notified_schools` variable has only `['Yes', 'No']` values



In [20]:
# Checking unique values in break_down_or_running_late variable to ensure there is no duplication
# ---
#
list(bus_df['breakdown_or_running_late'].unique())  

['Running Late', 'Breakdown']

In [21]:
# Checking unique values in school_age_or_prek variable
# ---
# 
list(bus_df['school_age_or_prek'].unique())  

['School-Age', 'Pre-K']

In [22]:
# Checking unique values in school_year variable
# ---
#  
list(bus_df['school_year'].unique())  

['2015-2016', '2017-2018', '2018-2019', '2016-2017', '2019-2020']

In [23]:
# Checking unique values in reason variable
# ---
#  
list(bus_df['reason'].unique())  

['Heavy Traffic',
 'Flat Tire',
 'Other',
 'Won`t Start',
 'Mechanical Problem',
 'Problem Run',
 'Accident',
 'Late return from Field Trip',
 'Delayed by School',
 'Weather Conditions',
 nan]

In [24]:
# Checking unique values in run_type variable
# ---
#  
list(bus_df['run_type'].unique())  

['Special Ed AM Run',
 'Pre-K/EI',
 'General Ed AM Run',
 'General Ed Field Trip',
 'Special Ed PM Run',
 'General Ed PM Run',
 'Project Read PM Run',
 'Special Ed Field Trip',
 'Project Read AM Run',
 'Project Read Field Trip',
 nan]

In [27]:
# Checking unique values in boro variable
# ---
#  
list(bus_df['boro'].unique())   

['New Jersey',
 'Manhattan',
 'Bronx',
 'Westchester',
 'Brooklyn',
 'Rockland County',
 'Nassau County',
 nan,
 'Queens',
 'Staten Island',
 'Connecticut',
 'All Boroughs']

### Overall Data Cleaning Observations
**Missing Values**

- There are a large number of missing values in the fields "How_Long_Delayed" which is important to our analysis. 
- There is an extremely large number of missing values in the "Incident_number" field but this is not incidental to our analysis and cannot be filled in without additional information.

**Error in values**

- "How_Long_Delayed" contains string values such as "MINS" or "mins" and a range of values, which needs to be changed to single integer value for our analysis.

**Error in Datatypes**

- "How_Long_Delayed" is a string datatype, should be converted to integer type.

**Error in field names**
- The column name "Boro" should be renamed to "Borough".


### Next Steps: Data Cleaning Steps

**Error in values**

- Extract the first integer value (lowest delay time) in the column "How_Long_Delayed"


**Missing Values**

- Impute the missing values in the field "How_Long_Delayed" with the mean value.


**Error in Datatypes**

- Convert "How_Long_Delayed" to int datatype.



**Error in field names**

- Rename the column "Boro" to "Borough".

In [40]:
# Lets first start by creating a copy of our dataframe 
# df_clean = df.copy(). We will use this copy as our cleaning copy.
# ---
#
df_clean = bus_df.copy()
df_clean.head()

school_year  busbreakdown_id           run_type bus_no route_number  \
0   2015-2016          1227538  Special Ed AM Run   2621         J711   
1   2015-2016          1227539  Special Ed AM Run   1260         M351   
2   2015-2016          1227540           Pre-K/EI    418            3   
3   2015-2016          1227541  Special Ed AM Run   4522         M271   
4   2015-2016          1227542  Special Ed AM Run   3124         M373   

          reason schools_serviced             occurred_on  \
0  Heavy Traffic            75003  11/05/2015 08:10:00 AM   
1  Heavy Traffic            06716  11/05/2015 08:10:00 AM   
2  Heavy Traffic             C445  11/05/2015 08:09:00 AM   
3  Heavy Traffic            02699  11/05/2015 08:12:00 AM   
4  Heavy Traffic            02116  11/05/2015 08:13:00 AM   

               created_on        boro  ... how_long_delayed  \
0  11/05/2015 08:12:00 AM  New Jersey  ...              NaN   
1  11/05/2015 08:12:00 AM   Manhattan  ...            20MNS   
2  11/05/2015 08:13:00 AM       Bronx  ...            15MIN   
3  11/05/2015 08:14:00 AM   Manhattan  ...           15 MIN   
4  11/05/2015 08:14:00 AM   Manhattan  ...              NaN   

  number_of_students_on_the_bus  has_contractor_notified_schools  \
0                            11                              Yes   
1                             2                              Yes   
2                             8                              Yes   
3                             6                               No   
4                             6                               No   

  has_contractor_notified_parents have_you_alerted_opt  \
0                              No                  Yes   
1                             Yes                   No   
2                             Yes                  Yes   
3                              No                   No   
4                              No                   No   

              informed_on incident_number         last_updated_on  \
0  11/05/2015 08:12:00 AM             NaN  11/05/2015 08:12:14 AM   
1  11/05/2015 08:12:00 AM             NaN  11/05/2015 08:13:34 AM   
2  11/05/2015 08:13:00 AM             NaN  11/05/2015 08:13:22 AM   
3  11/05/2015 08:14:00 AM             NaN  11/05/2015 08:14:04 AM   
4  11/05/2015 08:14:00 AM             NaN  11/05/2015 08:14:08 AM   

  breakdown_or_running_late school_age_or_prek  
0              Running Late         School-Age  
1              Running Late         School-Age  
2              Running Late              Pre-K  
3              Running Late         School-Age  
4              Running Late         School-Age  

[5 rows x 21 columns]

In [44]:
# Then extracting the lowest delay time in the column how_long_delayed from the string
# 
df_clean['how_long_delayed'] = df_clean['how_long_delayed'].str.extract('(\d+)')
df_clean['how_long_delayed'].head()

0    NaN
1     20
2     15
3     15
4    NaN
Name: how_long_delayed, dtype: object

We impute the null values in 'how_long_delayed' column with mean of the column. This will take a couple of steps...

In [45]:
# We first convert our how_long_delayed to float type to allow for imputation
# ---
# 
df_clean['how_long_delayed']= df_clean['how_long_delayed'].astype(float)


In [47]:
# Then later perform our mean imputation 
# ---
# 
df_clean['how_long_delayed'] = df_clean['how_long_delayed'].fillna(df_clean['how_long_delayed'].mean().round(4))

In [53]:
# Then convert back our how_long_delayed column to integer datatype
# ---
# 
df_clean['how_long_delayed']= df_clean['how_long_delayed'].apply(np.int64)
df_clean['how_long_delayed'].dtypes

dtype('int64')

In [51]:
# Then check for nulls in the column
# ---
# 
df_clean['how_long_delayed'].isna().sum()

0

In [58]:
# Rename Boro column to Borough
# ---
# 
df_clean.rename(columns = {'boro': 'Borough'}, inplace = True)


In [61]:
# Lastly we convert all values in our colums to lower case
# for ease of reading
# ---
# 
df_clean.columns.str.lower()

Index(['school_year', 'busbreakdown_id', 'run_type', 'bus_no', 'route_number',
       'reason', 'schools_serviced', 'occurred_on', 'created_on', 'borough',
       'bus_company_name', 'how_long_delayed', 'number_of_students_on_the_bus',
       'has_contractor_notified_schools', 'has_contractor_notified_parents',
       'have_you_alerted_opt', 'informed_on', 'incident_number',
       'last_updated_on', 'breakdown_or_running_late', 'school_age_or_prek'],
      dtype='object')

In [112]:
# Check the first 5 record the cleaned dataset
# ---
#
df_clean.head()

school_year  busbreakdown_id           run_type bus_no route_number  \
0   2015-2016          1227538  Special Ed AM Run   2621         J711   
1   2015-2016          1227539  Special Ed AM Run   1260         M351   
2   2015-2016          1227540           Pre-K/EI    418            3   
3   2015-2016          1227541  Special Ed AM Run   4522         M271   
4   2015-2016          1227542  Special Ed AM Run   3124         M373   

          reason schools_serviced             occurred_on  \
0  Heavy Traffic            75003  11/05/2015 08:10:00 AM   
1  Heavy Traffic            06716  11/05/2015 08:10:00 AM   
2  Heavy Traffic             C445  11/05/2015 08:09:00 AM   
3  Heavy Traffic            02699  11/05/2015 08:12:00 AM   
4  Heavy Traffic            02116  11/05/2015 08:13:00 AM   

               created_on     borough  ... how_long_delayed  \
0  11/05/2015 08:12:00 AM  New Jersey  ...               28   
1  11/05/2015 08:12:00 AM   Manhattan  ...               20   
2  11/05/2015 08:13:00 AM       Bronx  ...               15   
3  11/05/2015 08:14:00 AM   Manhattan  ...               15   
4  11/05/2015 08:14:00 AM   Manhattan  ...               28   

   number_of_students_on_the_bus  has_contractor_notified_schools  \
0                             11                              Yes   
1                              2                              Yes   
2                              8                              Yes   
3                              6                               No   
4                              6                               No   

  has_contractor_notified_parents have_you_alerted_opt  \
0                              No                  Yes   
1                             Yes                   No   
2                             Yes                  Yes   
3                              No                   No   
4                              No                   No   

              informed_on incident_number         last_updated_on  \
0  11/05/2015 08:12:00 AM             NaN  11/05/2015 08:12:14 AM   
1  11/05/2015 08:12:00 AM             NaN  11/05/2015 08:13:34 AM   
2  11/05/2015 08:13:00 AM             NaN  11/05/2015 08:13:22 AM   
3  11/05/2015 08:14:00 AM             NaN  11/05/2015 08:14:04 AM   
4  11/05/2015 08:14:00 AM             NaN  11/05/2015 08:14:08 AM   

  breakdown_or_running_late school_age_or_prek  
0              Running Late         School-Age  
1              Running Late         School-Age  
2              Running Late              Pre-K  
3              Running Late         School-Age  
4              Running Late         School-Age  

[5 rows x 21 columns]


## 5. Solution Implementation

Here we investigate the questions that would help craft our recommendations.

### 5.a) Questions

In [84]:
# 1. Which bus companies that had the highest breakdowns?
# ---
#
breakdowns = df_clean.groupby(['bus_company_name']).count()

# Sort to get bus company with highest breakdowns
# ---
# YOUR CODE GOES BELOW
#
breakdowns.sort_values(by='busbreakdown_id', ascending = False).head(1)

school_year  busbreakdown_id  run_type  bus_no  \
bus_company_name                                                   
G.V.C., LTD.            19394            19394     19394   19394   

                  route_number  reason  schools_serviced  occurred_on  \
bus_company_name                                                        
G.V.C., LTD.             19394   19394             19394        19394   

                  created_on  borough  how_long_delayed  \
bus_company_name                                          
G.V.C., LTD.           19394    19394             19394   

                  number_of_students_on_the_bus  \
bus_company_name                                  
G.V.C., LTD.                              19394   

                  has_contractor_notified_schools  \
bus_company_name                                    
G.V.C., LTD.                                19394   

                  has_contractor_notified_parents  have_you_alerted_opt  \
bus_company_name                                                          
G.V.C., LTD.                                19394                 19394   

                  informed_on  incident_number  last_updated_on  \
bus_company_name                                                  
G.V.C., LTD.            19394                0            19394   

                  breakdown_or_running_late  school_age_or_prek  
bus_company_name                                                 
G.V.C., LTD.                          19394               19394

In [83]:
# 2. What were the top 3 reasons for bus delays?
# ---
#
bus_delay = df_clean.groupby(['reason']).count()[['how_long_delayed']].reset_index()

# Sort to get the most frequent reason
# ---
# YOUR CODE GOES BELOW
#
bus_delay.sort_values(by ='how_long_delayed', ascending = False).head(3)

reason  how_long_delayed
3       Heavy Traffic            173221
6               Other             37579
5  Mechanical Problem             28162

In [148]:
# 3. How many students were in the buses when they broke down? 
# ---
#
df_clean.groupby(['number_of_students_on_the_bus']).count()[['busbreakdown_id']].reset_index()['number_of_students_on_the_bus'].sum()

310996

In [147]:
# 4. Which were most frequent reasons for bus breakdowns?
# ---
#
breakdown_reasons = df_clean.groupby(['reason']).count()

# Sort to get most frequent reasons
# ---
# YOUR CODE GOES BELOW
#
breakdown_reasons[['breakdown_or_running_late']].sort_values(by ='breakdown_or_running_late', ascending = False)

breakdown_or_running_late
reason                                                
Heavy Traffic                                   173221
Other                                            37579
Mechanical Problem                               28162
Won`t Start                                      12283
Flat Tire                                         8411
Weather Conditions                                6932
Late return from Field Trip                       5704
Problem Run                                       4068
Accident                                          2526
Delayed by School                                 2222

In [145]:
# 5. What were the most frequent reasons for the bus running late? 
# ---
#
reasons = df_clean.query("breakdown_or_running_late == 'Breakdown'").groupby(['reason']).count()
# Get the records with running late reasons and sort to get most frequent reasons
# ---
# YOUR CODE GOES BELOW
#
reasons[['breakdown_or_running_late']].sort_values(by ='breakdown_or_running_late', ascending = False)

breakdown_or_running_late
reason                                                
Mechanical Problem                               14985
Won`t Start                                       7731
Flat Tire                                         4038
Other                                             3530
Heavy Traffic                                      419
Accident                                           202
Weather Conditions                                  82
Late return from Field Trip                         35
Problem Run                                         28
Delayed by School                                    7

In [106]:
# 6. What was the average delay time for each reason type?
# ---
# 
avg_delay = df_clean.groupby('reason').mean().reset_index()

# Get the records with reasons and how long on average a delay took then sort 
# ---
# YOUR CODE GOES BELOW
#
avg_delay[['reason','how_long_delayed']].sort_values(by='how_long_delayed')

reason  how_long_delayed
1            Delayed by School         18.907741
3                Heavy Traffic         26.225550
6                        Other         28.421006
8           Weather Conditions         29.227351
2                    Flat Tire         29.392343
9                  Won`t Start         29.644061
5           Mechanical Problem         30.915205
7                  Problem Run         31.136676
0                     Accident         34.965162
4  Late return from Field Trip         62.235975

### 5.b) Recommendations

From the above analysis, below are our recommendations:

1.   Late returns from Field trip has a very high delay time, as such time management during field trips need to be considered.
2.   





## 6. Challenging your Solution

During this step, we review our solution and implement  approaches that could potentially provide a better outcome. In our case, we could propose the following question that wasn't answered in our solution because it couldn't have greatly contributed to our recommendation.

In [152]:
# Which boroughs experienced the most breakdowns?
# ---  
#
breakdowns = df_clean.groupby(['borough']).count()

# Sort to get bus company with highest breakdowns
# ---
# YOUR CODE GOES BELOW
#
breakdowns[['reason']].sort_values(by='reason',ascending =False)

reason
borough                
Bronx             70952
Manhattan         68177
Brooklyn          62239
Queens            38058
Staten Island     13907
Westchester        7093
Nassau County      4124
New Jersey         1583
Rockland County     875
All Boroughs        429
Connecticut         210

**Our observations:**

1.   Bronx experienced the highest breakdowns
2.   Conneccticut experience the lowest breakdowns

How does this observation tie to our solution?



## 7. Follow up questions

During this step, you rethink and propose other ways that you can improve your solution.

### a). Did we have the right data?

* Yes the data was enough to work with

### b). Do we need other data to answer our question?

You can look into the questions you brainstormed that you weren't taken into account during analysis due to a lack of data. Were those questions important to have been left out  of your analysis?

### c). Did we have the right question?

Were there any other questions that we needed to have answered?

* Yes, What was the relationship between reason for breakdown and boroughs.